File: LintCode/Medium/159. Find Minimum in Rotated Sorted Array/Solution.java
public class Solution {
    /**
     * @param nums: a rotated sorted array
     * @return: the minimum number in the array
     */
    public int findMin(int[] nums) {
        // write your code here
        
        if (nums.length == 0) return -1;
        
        int start = 0, end = nums.length - 1;
        int target = nums[end];
        
        while (start + 1 < end) {
            int mid = (end - start) / 2 + start;
            
            if (nums[mid] < target) {
                end = mid;
            } else {
                start = mid;
            }
        }
        
        if (nums[start] < target) {
            return nums[start];
        } else {
            return nums[end];
        }
        
        
    }
}

File: LintCode/Medium/159. Find Minimum in Rotated Sorted Array/Solution.py
class Solution:
    """
    @param nums: a rotated sorted array
    @return: the minimum number in the array
    """
    def findMin(self, nums):
        # write your code here
        start, end = 0, len(nums) - 1
        target = nums[-1]
        while (start + 1 < end):
            mid = (start + end) // 2

            if (nums[mid] > target):
                start = mid
            elif (nums[mid] < target):
                end = mid


        if nums[start] < nums[end]:
            return nums[start]
        else:
            return nums[end]
